giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+          }
+          if (extreme[4] != -1) {
+            coefficient[4] = extreme[4];
+          }
+          setIsValid = true;
+          return; // the finally-block will execute before return
+        }
+        // CHECKSTYLE: stop IllegalCatch
+      } catch (Exception e) {
+        // CHECKSTYLE: resume IllegalCatch
+        LOG.warn("addRecord: exception occurred!", e);
+      } finally {
+        // This inner try-finally block is necessary to ensure that the
+        // lock is always released.
+        try {
+          isValid = setIsValid;
+          printStats();
+        } finally {
+          lock.unlock();
+        }
+      }
+    }
+
+    /**
+     * Certain coefficients need to be within a specific range. For instance,
+     * If the coefficient is not in this range, we set it to the closest bound
+     * and re-run the linear regression.
+     *
+     * @param coefIndex Coefficient index
+     * @param lowerBound Lower bound
+     * @param upperBound Upper bound
+     * @param xValues double[][] matrix with data samples
+     * @param yValues double[] matrix with y samples
+     * @return True if coefficients were out-of-range
+     * @throws Exception
+     */
+    private boolean refineCoefficient(int coefIndex, double lowerBound,
+      double upperBound, double[][] xValues, double[] yValues)
+      throws Exception {
+
+      boolean result = false;
+      if (coefficient[coefIndex] < lowerBound ||
+        coefficient[coefIndex] > upperBound) {
+
+        double value;
+        if (coefficient[coefIndex] < lowerBound) {
+          value = lowerBound;
+        } else {
+          value = upperBound;
+        }
+        int ptr = -1;
+        if (validColumnIndices.size() >= 1 &&
+          validColumnIndices.get(validColumnIndices.size() - 1) == coefIndex) {
+          ptr = validColumnIndices.size() - 1;
+        } else if (validColumnIndices.size() >= 2 &&
+          validColumnIndices.get(validColumnIndices.size() - 2) == coefIndex) {
+          ptr = validColumnIndices.size() - 2;
+        }
+        if (ptr != -1) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("addRecord: coefficient at index " + coefIndex +
+              " is wrong in the regression, setting it to " + value);
+          }
+          // remove from valid column
+          validColumnIndices.remove(ptr);
+          // re-create the X matrix
+          fillXMatrix(dataSamples, validColumnIndices, xValues);
+          // adjust Y values
+          for (int i = 0; i < memorySamples.size(); ++i) {
+            yValues[i] -= value * dataSamples.get(i)[coefIndex];
+          }
+          // save new coefficient value in intermediate array
+          extreme[coefIndex] = value;
+          // re-run regression
+          mlr.newSampleData(yValues, xValues);
+          calculateRegression(coefficient, validColumnIndices, mlr);
+          result = true;
+        } else {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug(
```

Comment:
Essentially, the `refineCoefficient` method is called in a loop to set the coefficients that are out-of-bound to be in their bound. This approach is a viable approach as our method of finding linear regression is based on least square method which finds the minimum of a convex function. This means, if a coefficient is out of bound in a non-constrained regression calculation, setting it to the extremum value of its bound results in the minimum of the convex function in our constrained regression problem.

Since this method is called in a loop, there is a chance that the method is called with a coefficient that is already refined and is not in `validColumnIndices` anymore. In such case, the coefficient is set to an extreme of its bound, but later on it will be overridden by the values in `extreme` array.